### DISCUSSION TITLE
Wrong Audio Track

### DISCUSSION THREAD

**Discussion by robi2107** [aclap-dev/video-downloadhelper#985] [Help and Q&A]
### Your question

I am getting downloads with the wrong audio tracks off of YouTube in Firefox browser. The problem occurs where video has more than one audio track available. Like this one: https://www.youtube.com/watch?v=rVjY6jXOs2A where 10 audio tracks are available. Even though the first one on tle list is Russian and I've got English (original) language checked, the downloaded video is in Arab which is forth on the list. The problem persists no matter which format or video quality do I choose. How do I persuade the app to download the English audio track. I've checked the forum for similar topic but found no satisfying answer or it is too complicated.  Pls hlp.

### Addon & CoApp version and details

version: 9.0.2.12
target: mozilla
channel: stable
lang: sl
coapp: {"found":true,"path":"C:\\Program Files\\DownloadHelper CoApp\\vdhcoapp.exe","version":"2.0.19","new_version":false}
license: {"accepted":true,"email":"<removed>","key":"b98e78388ac44573"}
platform: x86-64 win
UA: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:127.0) Gecko/20100101 Firefox/127.0
{
    "valid_license_message_has_been_displayed": true,
    "download_directory": "C:\\Users\\Robi\\dwhelper",
    "last_download_directory": {
        "__serializer_tag": "option_some",
        "__serializer_value": {
            "__serializer_tag": "primitive"
        }
    },
    "never_show_no_incognito_msg_again": true,
    "view_options": {
        "all_tabs": true,
        "low_quality": false,
        "sort_by_status": true,
        "sort_reverse": false,
        "show_button_clean": true,
        "show_button_clean_all": false,
        "show_button_convert_local": false,
        "hide_downloaded": false
    },
    "open_count_store": 38,
    "successfull_dl": 12
}

**Comment by GrampaWildWilly** [comment]
I do not work for VDH.  I am just another user.

It's odd that you didn't find this discussion: #956.

Just wanted post this as a heads up.  I'm now going to investigate whether I can apply something resembling the approach I took in #956 to YouTube.  YouTube is just a rat's nest of exceptions & weird techniques.  I'm only about 50% optimistic I can figure out how to do this.  But I'm going to try.  Stay tuned.

**Reply by GrampaWildWilly** [reply]
My pessimism was well founded.  I used techniques I documented at great length in our old forum.  If you want to find that, it is a thread with the title `How to download from YouTube`.  You can try searching the old forum.  Or you can do a Google search (maybe other search engines will work as well but I tried it only with Google) with this search key

VDH How to download from YouTube

The first search result is a link to my tutorial.

I applied the approach you can read about in that thread.  I did discover audio tracks in both mp4 & webm format, low & medium quality, for languages such as Arabic, Spanish, Spanish (Latin American), Portuguese, Portuguese (Brazil), Hindi, Russian, French.  And oh yeah.  English.  I tried about 10 different URLs for audio tracks but none of them worked.  Entering the URLs into their own Firefox tab just gave me a blank screen.  No error.  Nothing.

So my technique looks promising but it doesn't work.  I don't know why the URLs aren't working.  We have encountered YouTube content (not recently) that seems to require some sort of special magic spell to make the URLs work.  It's some kind of odd authentication that YouTube requires on some of its content.  Some.  Not all.  But some.  And apparently, this Stalin documentary is among the some.

Best I can do.  Maybe @paulrouget Paul can solve this one.  I can't.

**Reply by GrampaWildWilly** [reply]
Just as an aside, the languages are not listed in the Network Monitor in the same order you are claiming they are listed.  Arabic is the first language.  English is the last.  I suspect that VDH is picking up the Arabic audio because it is the first one listed.  Not listed in the menu in the player, but listed in the Adaptive Format data in the Network Monitor.  I don't know what magic VDH is performing but the URLs I'm seeing in the Network Monitor just don't work for me.

**Reply by GrampaWildWilly** [reply]
But then there's Plan B.  I decided to look in the Hit Details provided by VDH.  I fished the HLS master manifest URL out of the Hit Details for the object labelled `#0 main url`.  Running that through ffprobe produced this:

[ffprobe.txt](https://github.com/user-attachments/files/16153161/ffprobe.txt)

Using information in that report, I used ffmpeg to download the 1440x1080 video with the English audio track.  As a separate step, I took advantage of the truly impressive array of subtitles being offered here.  There are many more languages covered in the captions than there are in the available audio tracks.  Here are my 2 downloads:

[Stalin Documentary mp4 Log Filtered.txt](https://github.com/user-attachments/files/16153517/Stalin.Documentary.mp4.Log.Filtered.txt)
[Stalin Documentary vtt Log Filtered.txt](https://github.com/user-attachments/files/16153568/Stalin.Documentary.vtt.Log.Filtered.txt)

Those are filtered logs.  My process automatically filters out lines that ffmpeg logs as `Skip` if there are enough such lines.  There were about 100 lines being skipped in each of these 2 downloads.  Trust me.  You're not missing anything salient.

This resulted in an MP4 that VLC was quite happy to play with English narration & English subtitles.

Near the bottom of each download log you will find a couple of lines reporting the size of the downloaded file, the duration of the download, & the speed of the download.  The captions, being a simple text file of only a few K, came in pretty quickly.  For the short time both those downloads were running -- I ran them concurrently -- each download was getting about 1.5 million bytes per second service according to the Resource Monitor.  Not terrible but certainly nothing to make you take notice.  But once the subtitles were done, the download of the video started speeding up.  You can see the average speed in the log.  A quite decent number.  But that's just an average.  The speed reached about 20 million bytes per second around the time the download completed.  I suspect it would have continued to speed up if it had run any longer.

If you are particularly sharp eyed you will notice that I used the ffmpeg supplied with our CoApp.  This tells me that there is a way for VDH to handle cases like this, including the captions.  How soon Paul will implement a case like this is a question only he can answer.  I suspect he won't answer because he doesn't want to commit to any timetable.  I suspect he's working on other things.  He has already promised us subtitle support, although without a date.  I would imagine that would show up in VDH before support for multiple language audio tracks.

So it's your choice.  You can wait for some unknown, probably long, time for VDH to handle this case.  Or you can crack the books & learn how to use ffmpeg yourself to get this content today, as I just did.  I mentioned another discussion above.  There is a link in there to my tutorial on ffmpeg in our old forum.  The tools are at your fingertips.  It's up to you whether you choose to learn how to use them.

**Reply by GrampaWildWilly** [reply]
At first glance, #973 doesn't look like a duplicate of this one.  But once I dug into it, it is amazingly similar to this one.

**Comment by Mattmjs** [comment]
That looks to be the unintended consequence of using hls streams for the downloads. I see this video is on Odysee and Rumble.
Go to one of those sites if you want it.
